src/cephadm/cephadm

```diff
@@ -162,15 +162,120 @@ class NFSGanesha(object):
     entrypoint = '/usr/bin/ganesha.nfsd'
     daemon_args = ['-F', '-L', 'STDERR']
 
+    def __init__(self,
+                 fsid,
+                 daemon_type,
+                 daemon_id,
+                 config_json):
+        # type: (str, str, Union[int, str], Dict) -> None
+        self.fsid = fsid
+        self.daemon_type = daemon_type
+        self.daemon_id = daemon_id
+
+        def json_get(key, default=None, require=False):
+            if require and not key in config_json.keys():
+                raise ValueError('{} missing from config-json'.format(key))
+            return config_json.get(key, default)
+
+        # rados config options
+        self.user = json_get('user', 'admin') # TODO: require?
+        self.pool = json_get('pool', require=True)
+        self.namespace = json_get('namespace')
+
+    @classmethod
+    def init(cls, fsid, daemon_type, daemon_id):
+        # type: (str, str, Union[int, str]) -> NFSGanesha
+        return cls(fsid, daemon_type, daemon_id, get_parm(args.config_json))
+
     @staticmethod
     def get_container_mounts(data_dir):
         # type: (str) -> Dict[str, str]
         mounts = dict()
         mounts[os.path.join(data_dir, 'config')] = '/etc/ceph/ceph.conf:z'
         # TODO: `ceph auth get-or-create` instead of admin keyring?
         mounts[os.path.join(data_dir, 'keyring')] = '/etc/ceph/keyring:z'
+        mounts[os.path.join(data_dir, 'ganesha.conf')] = '/etc/ganesha/ganesha.conf'
         return mounts
 
+    def get_daemon_name(self):
+        # type: () -> str
+        return '%s.%s' % (self.daemon_type, self.daemon_id)
+
+    def get_ganesha_config_object(self):
+        # type: () -> str
+        return 'conf-%s' % (self.get_daemon_name())
+
+    def get_rados_url(self):
+        # type: () -> str
+        url = 'rados://%s/' % (self.pool)
+        if self.namespace:
+            url += self.namespace + '/'
+        url += self.get_ganesha_config_object()
+        return url
+
+    def get_ganesha_config(self):
```

Comment:
IMO this should live in mgr/cephadm/module.py, not here, and be passed through as a file to simply write out... just like the monitoring components work.  With that change, I wonder if this NFSGanesha class is needed, since most of this logic would exist in mgr/cephadm instead.

Reply:
:+1: I'll rework this PR to move the ganesha config into the json file

Reply:
Reworked to pass `ganesha.conf` (and another additional /etc/ganesha/ conf file) via the `files` dict of the config-json